pallets/handles/src/lib.rs

```diff
@@ -0,0 +1,654 @@
+			let base_handle: Handle = handle.try_into().unwrap_or_default();
+			let base_handle_str = core::str::from_utf8(&base_handle).unwrap_or("");
+
+			let handle_converter = HandleConverter::new();
+			let canonical_handle_str = handle_converter.convert_to_canonical(&base_handle_str);
+			let canonical_handle_vec = canonical_handle_str.as_bytes().to_vec();
+			let canonical_handle: Handle = canonical_handle_vec.try_into().unwrap();
+			let suffix_index =
+				Self::get_next_suffix_index_for_canonical_handle(canonical_handle.clone())
+					.unwrap_or_default();
+
+			// Generate suffixes from the next available suffix index
+			let mut suffix_generator = SuffixGenerator::new(
+				T::HandleSuffixMin::get() as usize,
+				T::HandleSuffixMax::get() as usize,
+				&canonical_handle_str,
+			);
+			let sequence = suffix_generator.suffix_iter().enumerate();
+
+			for (i, suffix) in sequence {
+				if i >= suffix_index as usize && i < (suffix_index as usize + count as usize) {
+					suffixes.push(suffix as u16);
+				}
+				if suffixes.len() == count as usize {
+					break
+				}
+			}
+			let response =
+				PresumptiveSuffixesResponse { base_handle: base_handle.into(), suffixes };
+			response
+		}
+
+		/// Creates a full display handle by combining a base handle string with a suffix generated
+		/// from an index into the suffix sequence.
+		///
+		/// # Arguments
+		///
+		/// * `base_handle_str` - The base handle string.
+		/// * `suffix_sequence_index` - The index into the suffix sequence.
+		///
+		/// # Returns
+		///
+		/// * `Handle` - The full display handle.
+		///
+		pub fn create_full_handle_for_index(
+			base_handle_str: &str,
+			suffix_sequence_index: SequenceIndex,
+		) -> Handle {
+			// Convert base display handle into a canonical display handle
+			let handle_converter = HandleConverter::new();
+			let canonical_handle_str = handle_converter.convert_to_canonical(&base_handle_str);
+
+			// Generate suffix from index into the suffix sequence
+			let suffix = Self::generate_suffix_for_canonical_handle(
+				&canonical_handle_str,
+				suffix_sequence_index as usize,
+			);
+
+			let full_handle: Handle = Self::create_full_handle(base_handle_str, suffix);
+			full_handle
+		}
+
+		/// Creates a full display handle by combining a base handle string with supplied suffix
+		///
+		/// # Arguments
+		///
+		/// * `base_handle_str` - The base handle string.
+		/// * `suffix` - The numeric suffix .
+		///
+		/// # Returns
+		///
+		/// * `Handle` - The full display handle.
+		///
+		pub fn create_full_handle(base_handle_str: &str, suffix: HandleSuffix) -> Handle {
+			// Compose the full display handle from the base handle, "." delimeter and suffix
+			let mut full_handle_vec: Vec<u8> = vec![];
+			full_handle_vec.extend(base_handle_str.as_bytes());
+			full_handle_vec.extend(b"."); // The delimeter
+			let mut buff = [0u8; SUFFIX_MAX_DIGITS];
+			full_handle_vec.extend(suffix.numtoa(10, &mut buff)); // Use base 10
```

Comment:
The code already knows the length of the vector. With some reordering, we can use `with_capacity` and avoid reallocations when building the vector. Recommended by Rust Lang: [https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html#capacity-and-reallocation](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html)
```suggestion
                        let base_handle_bytes = base_handle_str.as_bytes();
                        let mut buff = [0u8; SUFFIX_MAX_DIGITS];
                        let suffix_str = suffix.numtoa(10, &mut buff); // Use base 10

                        let mut full_handle_vec = Vec::with_capacity(base_handle_bytes.len() + 1 + suffix_str.len());
                        full_handle_vec.extend(base_handle_bytes);
                        full_handle_vec.push(b'.');
                        full_handle_vec.extend(suffix_str);
```